devlib/instrument/gem5stats.py

```diff
@@ -0,0 +1,152 @@
+#    Copyright 2017 ARM Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from __future__ import division
+import csv
+import re
+
+from devlib.instrument import Instrument, CONTINUOUS, MeasurementsCsv
+from devlib.exception import TargetError, HostError
+
+class Gem5StatsInstrument(Instrument):
+    '''
+    Instrument enabling power monitoring in Gem5
+
+    This instrument parses the stats file produced by Gem5 and extracts power
+    values. More information on the "stats.txt" file structure can be found 
+    here:
+    
+    http://pages.cs.wisc.edu/~david/courses/cs752/Spring2015/gem5-tutorial/part1/gem5_stats.html
+    
+    This instrument also implements Gem5's ROIs (Regions of Interest). A ROI is
+    an annotation on a time interval marked by start and end cursors. This enables
+    power analysis on a specific time window which spans over the execution of
+    a workload for example. Up to 8 ROIs can be used, with or without 
+    overlapping.
+    '''
+
+    mode = CONTINUOUS
+    dump_head='---------- Begin Simulation Statistics ----------'
+    dump_tail='---------- End Simulation Statistics   ----------'
+    
+    def __init__(self, target, channel_map, stats_file_path):
+        super(Gem5StatsInstrument, self).__init__(target)
+        
+        # Duration since last dump is indexed by 'sim_seconds' in stats.txt
+        self.add_channel('Duration', 'time', 'sim_seconds')
+        # Track power only for now
+        for f in channel_map:
+            self.add_channel(f, 'power', channel_map[f])
+        
+        self._stats_file_path = stats_file_path
+        self._roi_state = [False] * 8
+        self.dumps = []
+
+    def start(self, roi=0):
```

Comment:
Hmm I see your point but I'm really not sure if marking and reading from ROIs should be split in two parts. ROIs seem to be perfectly appropriate to implement start() and stop() methods of this instrument in my opinion. 
What could be possible is to completely hide the notion of ROI to users (which would meet @setrofim's request to have API methods to match base signatures btw) and to assign a ROI to each instrument instance in order to differentiate their start/stop call from the unique stats.txt file. This could be done using a simple class counter. Does it sound reasonable ?